Mathlib/Tactic/MoveAdd.lean

```diff
@@ -0,0 +1,427 @@
+if sum.isAppOf op then
+  let inR := ← sum.getAppArgs.filterM fun r => do isDefEq R (← inferType r <|> pure R)
+  let new := ← inR.mapM (getAddends ·)
+  return new.foldl Array.append  #[]
+else return #[sum]
+
+variable (op : Name) in
+/-- Recursively compute the Array of `getAddends` Arrays by recursing into the expression `sum`
+looking for instance of the operation `op`.
+
+Possibly returns duplicates!
+-/
+partial def getOps (sum : Expr) : MetaM (Array ((Array Expr) × Expr)) := do
+let summands := ← getAddends op (← inferType sum <|> return sum) sum
+let (first, rest) := if summands.size == 1 then (#[], sum.getExprInputs) else
+  (#[(summands, sum)], summands)
+let rest := ← rest.mapM getOps
+return rest.foldl Array.append  first
+
+/-- `prepareOp sum` takes an `Expr`ession as input.  It assumes that `sum` is a well-formed
+term representing a repeated application of a binary operation and that the summands are the
+last two arguments passed to the operation.
+It returns the expression consisting of the operation with all its arguments already applied,
+except for the last two.
+This is similar to `Lean.Meta.mkAdd, Lean.Meta.mkMul`, except that the resulting operation is
+primed to work with operands of the same type as the ones already appearing in `sum`.
+
+This is useful to rearrange the operands.
+-/
+def prepareOp (sum : Expr) : Expr :=
+let opargs := sum.getAppArgs
+(opargs.toList.take (opargs.size - 2)).foldl (fun x y => Expr.app x y) sum.getAppFn
+
+/--  `sumList op exs` assumes that `op` is the `Name` of a binary operation.
+If `exs` is the list `[e₁, e₂, ..., eₙ]` of `Expr`essions, then `sumList` returns
+`op (op( ... op (op e₁ e₂) e₃) ... eₙ)`.
+-/
+partial
+def sumList1 (prepOp : Expr) : List Expr → Expr
+  | []    => default
+  | [a]    => a
+  | a::as => as.foldl (fun x y => Expr.app (prepOp.app x) y) a
+
+/-- `rankSums op tgt instructions` takes as input
+* the name `op` of a binary operation,
+* an `Expr`ession `tgt`,
+* a list `instructions` of pair `(expression, boolean)`.
+
+It extracts the maximal subexpressions of `tgt` whose head symbol is `op`
+(i.e. the maximal subexpressions that consist only of applications of the binary operation `op`),
+it rearranges the operands of such subexpressions following the order implied by `instructions`
+(as in `reorderUsing`),
+it returns the list of pairs of expressions `(old_sum, new_sum)`, for which `old_sum ≠ new_sum`
+sorted by decreasing value of `Lean.Expr.size`.
+In particular, a subexpression of an `old_sum` can only appear *after* its over-expression.
+-/
+def rankSums (op : Name) (tgt : Expr) (instructions : List (Expr × Bool)) :
+    MetaM (List (Expr × Expr)) := do
+  let sums := ← getOps op (← instantiateMVars tgt)
+  let candidates := ← sums.mapM fun (addends, sum) => do
+    let reord := reorderUsing addends.toList instructions
+    let resummed := sumList1 (prepareOp sum) reord
+    if (resummed != sum) then return some (sum, resummed) else return none
```

Comment:
I don't see any bind in this `do`, so this can be a normal `map`, right? (untested)
```suggestion
  let candidates := sums.map fun (addends, sum) =>
    let reord := reorderUsing addends.toList instructions
    let resummed := sumList1 (prepareOp sum) reord
    if (resummed != sum) then some (sum, resummed) else none
```